src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,506 @@
+sliceChunk pageNr n MCompactIndex{..} = do
+    cPrimary <- VU.freeze $ VUM.slice pageNr n mciPrimary
+    cClashes <- VU.freeze $ VUM.slice pageNr n mciClashes
+    pure Chunk{..}
+
+sliceChunkUnsafe :: Int -> Int -> MCompactIndex s k -> ST s (Chunk k)
+sliceChunkUnsafe pageNr n MCompactIndex{..} = do
+    cPrimary <- VU.unsafeFreeze $ VUM.slice pageNr n mciPrimary
+    cClashes <- VU.unsafeFreeze $ VUM.slice pageNr n mciClashes
+    pure Chunk{..}
+
+getFinalChunk :: MCompactIndex s k -> ST s (FinalChunk k)
+getFinalChunk mci@MCompactIndex{..} = do
+    fillRangeFinderToEnd mci
+
+    fcRangeFinder <- VU.freeze mciRangeFinder
+    fcTieBreaker <- readSTRef mciTieBreaker
+    pure FinalChunk{fcRangeFinderPrecision = mciRangeFinderPrecision, ..}
+
+getFinalChunkUnsafe :: MCompactIndex s k -> ST s (FinalChunk k)
+getFinalChunkUnsafe mci@MCompactIndex{..} = do
+    fillRangeFinderToEnd mci
+
+    fcRangeFinder <- VU.unsafeFreeze mciRangeFinder
+    fcTieBreaker <- readSTRef mciTieBreaker
+    pure FinalChunk{fcRangeFinderPrecision = mciRangeFinderPrecision, ..}
+
+{-------------------------------------------------------------------------------
+  SliceBits
+-------------------------------------------------------------------------------}
+
+class SliceBits a where
+  -- | @'sliceBits' offset size x@ slices from @x@ a string of bits of size
+  -- @n@ that starts at the 0-based @offset@.
+  --
+  -- Offsets are counted from the /top/, which means offset @0@ corresponds to
+  -- the most significant bit.
+  sliceBits :: Int -> Int -> a -> a
+
+  -- | Like 'sliceBits', but the offset is always @0@.
+  topBits :: Int -> a -> a
+
+-- | Like 'topBits', but converts the output to a 'Word16'.
+topBits16 :: (SliceBits a, Integral a) => Int -> a -> Word16
+topBits16 size x = fromIntegral $ topBits size x
+
+-- | Like 'sliceBits', but converts the output to a 'Word32'.
+sliceBits32 :: (SliceBits a, Integral a) => Int -> Int -> a -> Word32
+sliceBits32 offset size x = fromIntegral $ sliceBits offset size x
+
+-- | Useful for deriving-via a 'SliceBits' instance.
+newtype FiniteB a = FiniteB a
+
+instance FiniteBits a => SliceBits (FiniteB a) where
+  sliceBits :: Int -> Int -> FiniteB a -> FiniteB a
+  sliceBits offset size (FiniteB x) = FiniteB $ sliceBitsDefault offset size x
+
+  topBits :: Int -> FiniteB a -> FiniteB a
+  topBits size (FiniteB x) = FiniteB $ topBitsDefault size x
+
+sliceBitsDefault :: FiniteBits a => Int -> Int -> a -> a
+sliceBitsDefault offset size x = x `shiftL` offset `shiftR` (finiteBitSize x - size)
+
+topBitsDefault :: FiniteBits a => Int -> a -> a
+topBitsDefault size x = shiftR x (finiteBitSize x - size)
+
+deriving via FiniteB Word instance SliceBits Word
+deriving via FiniteB Word8 instance SliceBits Word8
+deriving via FiniteB Word16 instance SliceBits Word16
+deriving via FiniteB Word32 instance SliceBits Word32
+deriving via FiniteB Word64 instance SliceBits Word64
+deriving via FiniteB Int instance SliceBits Int
```

Comment:
I'm not convinced we need these instances. We should decide what type we're planning to use the compact index, user key type or already serialised.